### PR TITLE
add patch tests

### PR DESCRIPTION
**Pull request by kazupon**
ref: #2872

### CODE REVIEW

**Comment by Twiknight** [comment]
Thanks for the work.

I prefer moving `module` specialized cases to separated files ( attrs, classes, etc).

All `patch` need to do is to call their hooks at the right time.
A case suit for `patch` is not responsible for their correctness.

File: test/unit/modules/vdom/patch.spec.js [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
+import { createComponent } from 'core/vdom/create-component'
+import { createPatchFunction } from 'core/vdom/patch'
+import baseModules from 'core/vdom/modules/index'
+import * as nodeOps from 'web/runtime/node-ops'
+import platformModules from 'web/runtime/modules/index'
+import VNode from 'core/vdom/vnode'
+
+function prop (name) {
+  return obj => { return obj[name] }
+}
```
Comment (yyx990803) [inline test/unit/modules/vdom/patch.spec.js]: blank lines between functions please

**Comment by yyx990803** [comment]
Agree with @Twiknight that this file is a bit too big - let's split it into multiple files (similar to how snabbdom's tests are).

**Comment by kazupon** [comment]
Thanks review!
I'll try to split test cases into multiple files.

**Comment by kazupon** [comment]
Done

File: test/unit/modules/vdom/modules/transition.spec.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      const vnode1 = new VNode('div', {}, [
+        new VNode('p', { transition: { definition: 'expand', appear: true }},
+          undefined, undefined, undefined, undefined, vm
+        )
+      ])
+      let elm = patch(null, vnode1)
+      expect(elm.childNodes[0].classList.contains('expand-enter')).toBe(true)
+      // remove
+      const vnode2 = new VNode('div', {})
+      elm = patch(vnode1, vnode2)
```
Comment (yyx990803) [inline test/unit/modules/vdom/modules/transition.spec.js]: is this complete?

File: test/unit/modules/vdom/patch/hydration.spec.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    const node0 = createServerRenderedDOM()
+    const vnode1 = new VNode('div', {}, [
+      new VNode('span', {}),
+      new VNode('div', { hook: { init }}, [
+        new VNode('span', {}),
+        new VNode('span', {})
+      ])
+    ])
+    patch(node0, vnode1)
+    expect(result.length).toBe(1)
```
Comment (yyx990803) [inline test/unit/modules/vdom/patch/hydration.spec.js]: should also assert that each `VNode`'s `elm` property is correctly set to the corresponding DOM element.

**Pull request merged**